File: agents/planner-adk/app/schemas.py
"""Pydantic mirrors of the JSON Schemas in /contracts/schemas.

Kept intentionally close to the shared contracts so the Planner's HTTP API
stays schema-compatible with contracts/schemas/travel-request.schema.json
and travel-response.schema.json. Do not fork field names here without
updating /contracts and explaining the incompatibility (PROJECT_SPEC.md §13
and §42 rule 6).
"""
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


class TravelRequest(BaseModel):
    request_id: str | None = None
    origin: str
    destination: str
    start_date: str
    end_date: str
    travelers: int = Field(ge=1)
    budget: float = Field(ge=0)
    currency: str = "BRL"
    preferences: list[str] = Field(default_factory=list)


class SubResult(BaseModel):
    status: Literal["SUCCESS", "PARTIAL", "UNAVAILABLE", "UNKNOWN"] = "UNAVAILABLE"


class FlightResult(SubResult):
    options: list[dict] = Field(default_factory=list)
    recommended_option_id: str | None = None
    notes: str = ""


class HotelResult(SubResult):
    options: list[dict] = Field(default_factory=list)
    notes: str = ""


class ActivityResult(SubResult):
    days: list[dict] = Field(default_factory=list)
    notes: str = ""


class BudgetResult(BaseModel):
    status: Literal["SUCCESS", "PARTIAL", "UNAVAILABLE", "UNKNOWN"] = "UNKNOWN"
    budget_status: Literal["WITHIN_BUDGET", "NEAR_LIMIT", "OVER_BUDGET", "UNKNOWN"] = "UNKNOWN"
    flight_cost: float | None = None
    hotel_cost: float | None = None
    activity_cost: float | None = None
    food_estimate: float | None = None
    transport_estimate: float | None = None
    total: float = 0
    limit: float = 0
    remaining: float = 0
    notes: str = ""


class EnrichmentResult(BaseModel):
    status: Literal["SUCCESS", "SKIPPED", "UNAVAILABLE"] = "SKIPPED"
    provider: str | None = None


class TravelResponseMetadata(BaseModel):
    trace_id: str = ""
    correlation_id: str = ""
    duration_ms: float = 0


class TravelResponse(BaseModel):
    request_id: str
    status: Literal["COMPLETED", "PARTIAL", "FAILED"]
    flight: FlightResult
    hotel: HotelResult
    activities: ActivityResult
    budget: BudgetResult
    enrichment: EnrichmentResult
    metadata: TravelResponseMetadata

File: agents/hotel-langgraph/src/schemas.ts
import { z } from "zod";

// Mirrors contracts/schemas/travel-request.schema.json (the subset the
// Hotel Agent needs). Kept intentionally close to /contracts — do not
// fork field names without updating the shared schema and explaining the
// incompatibility (PROJECT_SPEC.md §42 rule 6).
// A non-ISO date (e.g. "20/09/2026") used to pass `z.string().min(1)`
// silently, then turned into `Date.parse(...) === NaN` downstream in
// filter.ts — which made every hotel fail the budget filter and get
// reported as "no hotels matched the given criteria" (a validation bug
// disguised as a legitimate empty result). Reject it at parse time
// instead.
const isoDate = z.string().regex(/^\d{4}-\d{2}-\d{2}$/, "must be an ISO-8601 date (YYYY-MM-DD)");

export const HotelSearchRequestSchema = z.object({
  destination: z.string().min(1),
  start_date: isoDate,
  end_date: isoDate,
  travelers: z.number().int().min(1).default(1),
  budget: z.number().min(0).optional(),
  currency: z.string().optional(),
  preferences: z.array(z.string()).optional(),
});
export type HotelSearchRequest = z.infer<typeof HotelSearchRequestSchema>;

export const HotelOptionSchema = z.object({
  id: z.string(),
  name: z.string(),
  price_per_night: z.number(),
  currency: z.string(),
  rating: z.number().optional(),
  location: z.string().optional(),
});
export type HotelOption = z.infer<typeof HotelOptionSchema>;

// Mirrors contracts/schemas/hotel-result.schema.json.
export const HotelResultSchema = z.object({
  status: z.enum(["SUCCESS", "PARTIAL", "UNAVAILABLE", "UNKNOWN"]),
  options: z.array(HotelOptionSchema),
  notes: z.string().optional(),
});
export type HotelResult = z.infer<typeof HotelResultSchema>;
